// eslint-etc/dist/cjs/create-rule-tester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRuleTester = void 0;
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const path_1 = require("path");
function createRuleTester({ filename = (0, path_1.resolve)("./tests/file.ts"), parser = (0, path_1.resolve)("./node_modules/@typescript-eslint/parser"), project = (0, path_1.resolve)("./tests/tsconfig.json"), } = {}) {
    return function ruleTester({ comments = false, typeScript = true, types = true, } = {}) {
        const parserOptions = {
            comments,
            ecmaFeatures: { jsx: true },
            ecmaVersion: 2020,
            project: typeScript && types ? project : undefined,
            sourceType: "module",
        };
        const tester = new experimental_utils_1.TSESLint.RuleTester({
            parser: typeScript ? parser : undefined,
            parserOptions,
        });
        const run = tester.run;
        tester.run = (name, rule, { invalid = [], valid = [] }) => run.call(tester, name, rule, {
            invalid: invalid.map((test) => ({ ...test, filename })),
            valid: valid.map((test) => typeof test === "string"
                ? { code: test, filename }
                : { ...test, filename }),
        });
        return tester;
    };
}
exports.createRuleTester = createRuleTester;
