// eslint-plugin-lit/lib/rules/no-invalid-html.js
"use strict";
/**
 * @fileoverview Disallows invalid HTML in templates
 * @author James Garbutt <https://github.com/43081j>
 */
const template_analyzer_1 = require("../template-analyzer");
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
    meta: {
        docs: {
            description: 'Disallows invalid HTML in templates',
            category: 'Best Practices',
            url: 'https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/no-invalid-html.md'
        },
        messages: {
            parseError: 'Template contained invalid HTML syntax, error was: {{ err }}'
        }
    },
    create(context) {
        const source = context.getSourceCode();
        //----------------------------------------------------------------------
        // Helpers
        //----------------------------------------------------------------------
        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------
        return {
            TaggedTemplateExpression: (node) => {
                if (node.type === 'TaggedTemplateExpression' &&
                    node.tag.type === 'Identifier' &&
                    node.tag.name === 'html') {
                    const analyzer = template_analyzer_1.TemplateAnalyzer.create(node);
                    for (const err of analyzer.errors) {
                        // Ignore these as the duplicate attributes rule handles them
                        if (err.code === 'duplicate-attribute') {
                            continue;
                        }
                        const loc = analyzer.resolveLocation(err, source);
                        if (loc) {
                            context.report({
                                loc: loc,
                                messageId: 'parseError',
                                data: { err: err.code }
                            });
                        }
                    }
                }
            }
        };
    }
};
module.exports = rule;
