// chromium-bidi/lib/esm/protocol-parser/generated/webdriver-bidi-nav-speculation.js
/**
 * Copyright 2024 Google LLC.
 * Copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * THIS FILE IS AUTOGENERATED by cddlconv 0.1.7.
 * Run `node tools/generate-bidi-types.mjs` to regenerate.
 * @see https://github.com/w3c/webdriver-bidi/blob/master/index.bs
 */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck Some types may be circular.
import z from 'zod';
export var Speculation;
(function (Speculation) {
    Speculation.PreloadingStatusSchema = z.lazy(() => z.enum(['pending', 'ready', 'success', 'failure']));
})(Speculation || (Speculation = {}));
export const SpeculationEventSchema = z.lazy(() => Speculation.PrefetchStatusUpdatedSchema);
(function (Speculation) {
    Speculation.PrefetchStatusUpdatedSchema = z.lazy(() => z.object({
        method: z.literal('speculation.prefetchStatusUpdated'),
        params: Speculation.PrefetchStatusUpdatedParametersSchema,
    }));
})(Speculation || (Speculation = {}));
(function (Speculation) {
    Speculation.PrefetchStatusUpdatedParametersSchema = z.lazy(() => z.object({
        context: z.string(),
        url: z.string(),
        status: Speculation.PreloadingStatusSchema,
    }));
})(Speculation || (Speculation = {}));
